run_southafrica.py

```diff
@@ -1,95 +1,129 @@
+
+    args = parser.parse_args()
+    return args
+
+
+def main(args):
+    """Main entry point allowing external calls
+
+    Args:
+      args ([str]): command line parameter list
+    """
+    args = parse_args(args)
+    if args.path is not None:
+        download_path = str(args.path)
     else:
-        new_za.delete_unneeded_files('duplicates-nomatch-'+specs,[],files_path=new_za.downloadPath+'/mandates')
-    new_za.delete_unneeded_files('duplicates-'+specs,[])
-    new_za.teardown()
-elif args.webpage == 3: 
-    matches_acts = new_za.run_acts(keywords)
-    specs = 'SouthAfrica-acts'
-    if len(matches_acts) > 0:
-        print(matches_acts) 
-        new_za.print_matches(matches_acts,specs)
-        new_za.delete_no_matches(specs,path=new_za.downloadPath+'/acts')
+        current_dir = Path(os.path.abspath(os.curdir))
+        download_path = current_dir / 'south_africa_output'
+        download_path = str(download_path)  # class wants strings for paths
+        if not os.path.exists(download_path):
+            os.mkdir(download_path)
+
+    # setting up options for Chromedriver
+    # mostly to ensure any PDF links automatically
+    # download the PDFs to download_path
+    options = webdriver.ChromeOptions()
+    options.add_experimental_option('prefs', {
+        "download.default_directory": download_path,  # Default dir downloads
+        "download.prompt_for_download": False,  # to auto download the file
+        "download.directory_upgrade": True,
+        "plugins.always_open_pdf_externally": True  # don't show PDF in chrome
+    })
+
+    pd.options.display.max_colwidth = 1000
+    keywords = ['legal aid']  # TODO: take these as command-line options?
+
+    # keywords = ['legal aid',
+    #           'legal assistance',
+    #           'legal service',
+    #           'judicial assistance']
+
+    new_za = legisSouthAfrica(args.driver, download_path, options)
+    new_za.checkers()
+
+    if args.webpage == 1:
+        scrape(new_za, keywords, 'constit')
+    elif args.webpage == 2:
+        scrape(new_za, keywords, 'mandates')
+    elif args.webpage == 3:
+        scrape(new_za, keywords, 'acts')
+    elif args.webpage == 4:
+        scrape(new_za, keywords, 'bills')
     else:
-        new_za.delete_unneeded_files('duplicates-nomatch-'+specs,[],files_path=new_za.downloadPath+'/acts')
-    new_za.delete_unneeded_files('duplicates-'+specs,[])
-    new_za.teardown()
-elif args.webpage == 4: 
-    matches_bills = new_za.run_bills(keywords)
-    specs = 'SouthAfrica-bills'
-    if len(matches_bills) > 0:
-        print(matches_bills) 
-        new_za.print_matches(matches_bills,specs)
-        new_za.delete_no_matches(specs,path=new_za.downloadPath+'/bills')
+        error_msg = 'ERROR: webpage integer indicator not found.'
+        raise ValueError(error_msg + webpage_help)
+
+
+def scrape(new_za, keywords: list, page_type: str):
+    matches = new_za.run_bills(keywords)
```

Comment:
this is the biggie. In the `legisSouthAfrica` class I've created a few different `run_*()` functions based on the `page_type`, so there would need to be a set of conditionals within `scrape()` to determine which one to use.

Reply:
ah yes, right, the function would depend on the `page_type`. sloppy mistake on my part. Is there a method like `run` that takes the type of run as an argument and calls the appropriate `run_X` function?

something like

```python
def run(page_type, keywords):
    if page_type == 'bills':
        return run_bills(keywords)
    ...
```

Reply:
if so, that would allow the `scrape` function to generalize nicely.

Reply:
there isn't, but I can write one if you want (or you can have a crack at it if you want!). I think something like what you've written would work as a method in the `legisSouthAfrica` class (that's currently the only class that has this weird structure).